### PR TITLE
Avoid accessing GAC when not available (on CoreCLR)

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

**Comment by tmat** [comment]
@agocke @jaredpar

**Comment by tmat** [comment]
Fixing up ETA now.

**Comment by agocke** [comment]
I like your shim additions. I'll pull into my branch.

File: src/Compilers/Helpers/CoreClrShim.cs [hunk 1]
```diff
@@ -33,17 +29,22 @@ private static void Touch(Type type)
...
         internal static class CodePagesEncodingProvider
         {
             internal static readonly Type Type =
-                Type.GetType("System.Text.CodePagesEncodingProvider, System.Text.Encoding.CodePages, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a",
-                             throwOnError: false);
+                ReflectionUtilities.TryGetType("System.Text.CodePagesEncodingProvider, System.Text.Encoding.CodePages, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a");
 
-            private static PropertyInfo s_instance = Type
-                .GetTypeInfo()
+            private static PropertyInfo s_instance = Type?
```
Comment (jaredpar) [inline src/Compilers/Helpers/CoreClrShim.cs]: space before the ?

**Comment by tmat** [comment]
FWIW i think the shims should be all lazy, not based on type loader lazy loading but explicitly lazy (e.g. Lazy<T>). It doesn't seem right to upfront load all these types thru reflection that might not ever be needed.

**Comment by tmat** [comment]
test win please

**Comment by tmat** [comment]
test eta please

File: src/Compilers/Helpers/GlobalAssemblyCacheHelpers/GacFileResolver.cs [hunk 2]
```diff
@@ -11,37 +11,46 @@
...
+    /// Resolves assembly identities in Global Assembly Cache.
     /// </summary>
     internal sealed class GacFileResolver : IEquatable<GacFileResolver>
     {
         /// <summary>
-        /// Architecture filter used when resolving assembly references.
+        /// Returns true if GAC is available on the platform.
         /// </summary>
-        public ImmutableArray<ProcessorArchitecture> Architectures { get; }
+        public static bool IsAvailable => CoreClrShim.AssemblyLoadContext.Type == null;
```
Comment (jaredpar) [inline src/Compilers/Helpers/GlobalAssemblyCacheHelpers/GacFileResolver.cs]: This is a bit indirect.  I'd prefer we have a property somewhere which directly declared: IsRuntimeCoreClr, IsRuntimeDesktop, etc ...

**Comment by tmat** [comment]
test mac please

File: src/Compilers/Helpers/GlobalAssemblyCacheHelpers/GacFileResolver.cs (hunk 2 above)
Comment (tmat) [inline src/Compilers/Helpers/GlobalAssemblyCacheHelpers/GacFileResolver.cs]: In general we should be asking whether a specific feature is available, not on which platform we are. 

It's true that in this case an absence of AssemblyLoadContext doesn't neccessarily mean presence of GAC, so I guess a better way to check would be to try GetProcAddress on the clr.dll entry points that implement factories for GAC COM interfaces. I can follow up on that later. For now check for AssemblyLoadContext is close enough.

**Comment by tmat** [comment]
test mac please

File: src/Compilers/Helpers/GlobalAssemblyCacheHelpers/GacFileResolver.cs (hunk 2 above)
Comment (jaredpar) [inline src/Compilers/Helpers/GlobalAssemblyCacheHelpers/GacFileResolver.cs]: Do we have a bug tracking this change?

**Comment by tmat** [comment]
test mac please

**Comment by agocke** [comment]
:+1: As long as a bug is filed for fixing up the GAC dynamic checking later.

**Comment by jaredpar** [comment]
test mac please

**Comment by jaredpar** [comment]
@mmitche took the bad mac off line (suffering from a hard drive issue)

**Comment by mmitche** [comment]
I wonder why this isn't retriggering...all the rest seem to be triggering fine

**Comment by mmitche** [comment]
test mac please

File: src/Compilers/Helpers/GlobalAssemblyCacheHelpers/GacFileResolver.cs (hunk 2 above)
Comment (tmat) [inline src/Compilers/Helpers/GlobalAssemblyCacheHelpers/GacFileResolver.cs]: https://github.com/dotnet/roslyn/issues/5538

**Pull request merged**